pandas/core/groupby/ops.py

```diff
@@ -574,9 +587,52 @@ def _ea_wrap_cython_operation(
             f"function is not implemented for this dtype: {values.dtype}"
         )
 
+    @final
+    def _masked_ea_wrap_cython_operation(
+        self,
+        kind: str,
+        values: BaseMaskedArray,
+        how: str,
+        axis: int,
+        min_count: int = -1,
+        **kwargs,
+    ) -> BaseMaskedArray:
+        """
+        Equivalent of `_ea_wrap_cython_operation`, but optimized for masked EA's
+        and cython algorithms which accept a mask.
+        """
+        orig_values = values
+
+        # isna just directly returns self._mask, so copy here to prevent
+        # modifying the original
+        mask = isna(values).copy()
+        arr = values._data
```

Comment:
can we avoid depending on the MaskedArray implementation details?

Reply:
Avoiding the private attribute would be nice - best alternative I could think of is something like
`arr = arr.to_numpy(na_value=0, dtype=arr.dtype.type)`. But this seems a bit unintuitive since it fills masked positions whose value we don't actually care about.

Reply:
i think the thing to do is implement the appropriate method on the EA directly, something like `ordinal_operation`

(side-note: im unclear on how we handle np.nan vs pd.NA for FloatingArray)

Reply:
I think it is fine to depend on the implementation details here (we actually *want* to depend on this to have efficient implementations, we just didn't yet officially expose this, see also https://github.com/pandas-dev/pandas/issues/34873, where we basically said it's OK for now to use `_data`/`_mask` on your own risk)

Reply:
i guess nothing prevents us from later moving this to the relevant EA subclass if we figure out a general-case method for it to go under.

@jorisvandenbossche any thoughts on the np.nan vs pd.NA thing?

Reply:
Been some discussion primarily in #32265, also popping up various places like #39926, #30188

Reply:
> any thoughts on the np.nan vs pd.NA thing?

Since the general discussion is not yet resolved, I would simply not deal with NaN specifically for now (as the PR does right now, it only checks the mask if one is passed).